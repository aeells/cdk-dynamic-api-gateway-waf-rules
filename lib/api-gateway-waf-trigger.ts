import {Construct} from 'constructs';

import * as triggers from 'aws-cdk-lib/triggers';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as path from 'path';

export interface ApiGatewayWafTriggerProps
{
    readonly restApiId: string;
    readonly stageName: string;
}

export class ApiGatewayWafTrigger extends Construct
{
    readonly triggerFunction: triggers.TriggerFunction;

    constructor(scope: Construct, id: string, properties: ApiGatewayWafTriggerProps)
    {
        super(scope, id);

        this.triggerFunction = new triggers.TriggerFunction(this, 'ApiGatewayOpenApiPutFn', {
            runtime: lambda.Runtime.NODEJS_16_X,
            handler: 'index.handler',
            code: lambda.Code.fromAsset(path.join(__dirname, 'waf')),
            environment: {
              REST_API_ID: properties.restApiId,
              STAGE_NAME: properties.stageName
            },
            initialPolicy: [
                new iam.PolicyStatement({
                    effect: iam.Effect.ALLOW,
                    actions: ['apigateway:GET', 'ssm:*'],
                    resources: ['*']
                })
            ]
        })
    }
}