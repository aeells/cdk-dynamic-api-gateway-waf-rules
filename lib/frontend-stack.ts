import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as waf from 'aws-cdk-lib/aws-wafv2';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { Fn } from 'aws-cdk-lib';

export interface WafRule
{
    name: string;
    rule: waf.CfnWebACL.RuleProperty;
}

export class FrontEndStack extends cdk.Stack
{
    readonly openApiPaths: string[];

    constructor(scope: Construct, id: string, props?: cdk.StackProps)
    {
        super(scope, id, props);

        this.openApiPaths = Fn.split(',',
            ssm.StringParameter.fromStringParameterAttributes(this, 'ERSchedulerSSMOpenApiPaths', {
                parameterName: 'openapi-paths',
            }).stringValue
        );

        this.createWafAcl();
    }

    private createWafAcl() {
        const wafRules: WafRule[] = [
            ...this.getApiGatewayPathAllowRules(10), // 10+ Allow
            this.getApiGatewayPathBlockRule(20), // 20 Block
        ];

        return new waf.CfnWebACL(this, 'WafV2', {
            defaultAction: { allow: {} },
            visibilityConfig: {
                cloudWatchMetricsEnabled: false,
                metricName: 'cloudwatch-metrics-disabled-for-example-code',
                sampledRequestsEnabled: false,
            },
            scope: 'REGIONAL',
            rules: wafRules.map((wafRule) => wafRule.rule),
        });
    }

    // todo aeells - it might be more efficient to create a single rule with multiple statements?
    private getApiGatewayPathAllowRules(priority: number): WafRule[] {
        const rules: WafRule[] = [];
        for (let i = 0; i < this.openApiPaths.length; i++) {
            const apiPath: string = Fn.select(i, this.openApiPaths);
            rules.push(this.createApiGatewayRegexAllowRule(apiPath, priority++));
        }

        return rules;
    }

    private createApiGatewayRegexAllowRule(path: string, priority: number): WafRule {
        return {
            name: `ApiGatewayPathAllows${priority}`,
            rule: {
                name: `ApiGatewayPathAllows${priority}`,
                priority: priority,
                action: {
                    allow: {},
                },
                visibilityConfig: {
                    sampledRequestsEnabled: true,
                    cloudWatchMetricsEnabled: true,
                    metricName: 'ApiGatewayPathAllows',
                },
                statement: {
                    regexMatchStatement: {
                        fieldToMatch: {
                            uriPath: {},
                        },
                        regexString: path,
                        textTransformations: [
                            {
                                type: 'NONE',
                                priority: 0,
                            },
                        ],
                    },
                },
            },
        };
    }

    private getApiGatewayPathBlockRule(priority: number): WafRule {
        return {
            name: 'ApiGatewayPathBlocks',
            rule: {
                name: 'ApiGatewayPathBlocks',
                priority: priority,
                action: {
                    block: {},
                },
                visibilityConfig: {
                    sampledRequestsEnabled: true,
                    cloudWatchMetricsEnabled: true,
                    metricName: 'ApiGatewayPathBlocks',
                },
                statement: {
                    byteMatchStatement: {
                        fieldToMatch: {
                            uriPath: {},
                        },
                        positionalConstraint: 'STARTS_WITH',
                        searchString: '/api/',
                        textTransformations: [
                            {
                                type: 'NONE',
                                priority: 0,
                            },
                        ],
                    },
                },
            },
        };
    }
}
