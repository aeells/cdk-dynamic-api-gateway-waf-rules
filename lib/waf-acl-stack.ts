import {Construct} from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as waf from 'aws-cdk-lib/aws-wafv2';
import * as ssm from 'aws-cdk-lib/aws-ssm';

export interface WafRule
{
    name: string;
    rule: waf.CfnWebACL.RuleProperty;
}

export class WafAclStack extends cdk.Stack
{
    private openApiPaths: string[];

    constructor(scope: Construct, id: string, props?: cdk.StackProps)
    {
        super(scope, id, props);

        // retrieve the REST API endpoint paths from SSM
        this.retrieveRestApiEndpointPaths();

        // create WAF Web ACL rules locked down to those endpoints
        this.createWafAcl();
    }

    private retrieveRestApiEndpointPaths()
    {
        this.openApiPaths = cdk.Fn.split(',',
            ssm.StringParameter.fromStringParameterAttributes(this, 'SSMOpenApiPaths', {
                parameterName: 'api-paths',
            }).stringValue,
            // todo aeells - array 'assumedLength' is a restriction of the CDK and is required here to iterate properly
            // restriction documented in full in the Fn#split api https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-split.html

            // if your API is under active development and regularly  changing in size it would
            // be fairly simple to add a second 'array length' parameter to the Parameter Store
            2
        );
    }

    private createWafAcl() {
        const wafRules: WafRule[] = [
            ...this.getApiGatewayPathAllowRules(10), // 10+ Allow
            this.getApiGatewayPathBlockRule(20), // 20 Block
        ];

        return new waf.CfnWebACL(this, 'WafV2', {
            defaultAction: { allow: {} },
            visibilityConfig: {
                cloudWatchMetricsEnabled: false,
                metricName: 'cloudwatch-metrics-disabled-for-example-code',
                sampledRequestsEnabled: false,
            },
            scope: 'REGIONAL',
            rules: wafRules.map((wafRule) => wafRule.rule),
        });
    }

    private getApiGatewayPathAllowRules(priority: number): WafRule[] {
        const rules: WafRule[] = [];
        for (let i = 0; i < this.openApiPaths.length; i++) {
            const apiPath: string = cdk.Fn.select(i, this.openApiPaths);
            rules.push(this.createApiGatewayRegexAllowRule(apiPath, priority++));
        }

        return rules;
    }

    private createApiGatewayRegexAllowRule(path: string, priority: number): WafRule {
        return {
            name: `ApiGatewayPathAllows${priority}`,
            rule: {
                name: `ApiGatewayPathAllows${priority}`,
                priority: priority,
                action: {
                    allow: {},
                },
                visibilityConfig: {
                    sampledRequestsEnabled: true,
                    cloudWatchMetricsEnabled: true,
                    metricName: 'ApiGatewayPathAllows',
                },
                statement: {
                    regexMatchStatement: {
                        fieldToMatch: {
                            uriPath: {},
                        },
                        regexString: path,
                        textTransformations: [
                            {
                                type: 'NONE',
                                priority: 0,
                            },
                        ],
                    },
                },
            },
        };
    }

    private getApiGatewayPathBlockRule(priority: number): WafRule {
        return {
            name: 'ApiGatewayPathBlocks',
            rule: {
                name: 'ApiGatewayPathBlocks',
                priority: priority,
                action: {
                    block: {},
                },
                visibilityConfig: {
                    sampledRequestsEnabled: true,
                    cloudWatchMetricsEnabled: true,
                    metricName: 'ApiGatewayPathBlocks',
                },
                statement: {
                    byteMatchStatement: {
                        fieldToMatch: {
                            uriPath: {},
                        },
                        positionalConstraint: 'STARTS_WITH',
                        searchString: '/',
                        textTransformations: [
                            {
                                type: 'NONE',
                                priority: 0,
                            },
                        ],
                    },
                },
            },
        };
    }
}
