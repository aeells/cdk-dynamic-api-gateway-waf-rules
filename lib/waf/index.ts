import * as AWS from 'aws-sdk';

export const handler = async (
    event:  any
) => {
    const apiExport : any = await new AWS.APIGateway().getExport({
        exportType: 'oas30',
        restApiId: `${process.env.REST_API_ID}`,
        stageName: `${process.env.STAGE_NAME}`,
        accepts: 'application/json'
    }).promise();

    const apiPaths: string[] = [];
    for (const [key] of Object.entries(JSON.parse(apiExport.body.toString()).paths))
    {
        apiPaths.push(key.replace('{bookId}', '[A-Za-z0-9]+-[A-za-z0-9]+'));
    }

    await new AWS.SSM().putParameter({
        Name: 'api-paths',
        Value: apiPaths.toString(),
        Overwrite: true,
        Type: 'String'
    }).promise();
};

