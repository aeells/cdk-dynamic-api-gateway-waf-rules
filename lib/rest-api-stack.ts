import {Construct} from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import { ApiGatewayWafTrigger } from "./api-gateway-waf-trigger";

export class RestApiStack extends cdk.Stack
{
    private restApi: apigateway.RestApi;
    
    constructor(scope: Construct, id: string, props?: cdk.StackProps)
    {
        super(scope, id, props);

        // in the real world, our api endpoint resources are loaded from an external package from another development team
        this.createDummyRestApi();

        // there is no cdk api to list/get RestApi resources so using a lambda and sdk
        this.createApiGatewayWafTrigger();
    }

    private createDummyRestApi()
    {
        this.restApi = new apigateway.RestApi(this, 'ApiGatewayRestApi', {
            description: 'example api gateway',
        });

        const books = this.restApi.root.addResource('books');
        books.addMethod('GET');
        books.addMethod('POST');

        const book = books.addResource('{bookId}');
        book.addMethod('GET');
        book.addMethod('DELETE');

        // ...etc...
    }

    private createApiGatewayWafTrigger()
    {
        const apiGatewayWafTrigger = new ApiGatewayWafTrigger(this, 'ApiGatewayWafTrigger', {
            restApiId: this.restApi.restApiId,
            stageName: 'prod',
        });

        apiGatewayWafTrigger.triggerFunction.executeAfter(this.restApi);
    }
}
