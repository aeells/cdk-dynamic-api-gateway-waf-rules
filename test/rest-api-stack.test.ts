import { Template } from 'aws-cdk-lib/assertions';
import { App } from 'aws-cdk-lib';
import {RestApiStack} from '../lib/rest-api-stack';

describe('REST API stack test', () => {
    const app = new App();

    const restApiStack = new RestApiStack(app, 'RestApiStack', {
        env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
    });

    test('an API Gateway REST API exists with Lambda and associated TriggerFunction as per source-controlled snapshot', () => {
        expect(Template.fromStack(restApiStack)).toMatchSnapshot();
    });
});
