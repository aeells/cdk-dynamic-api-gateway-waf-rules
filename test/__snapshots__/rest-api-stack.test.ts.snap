// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REST API stack test an API Gateway REST API exists with Lambda and associated TriggerFunction as per source-controlled snapshot 1`] = `
{
  "Outputs": {
    "ApiGatewayRestApiEndpointB0E85E39": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "ApiGatewayRestApi4735CE82",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "ApiGatewayRestApiDeploymentStageprodFFF8B293",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91": {
      "DependsOn": [
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b5b182c36096f19283e1157c12aed83c585041a5e7c0547e8fe9b548bd22ebf2.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          {
                            "Fn::GetAtt": [
                              "ApiGatewayWafTriggerApiGatewayOpenApiPutFn1601EB8B",
                              "Arn",
                            ],
                          },
                          ":*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiGatewayRestApi4735CE82": {
      "Properties": {
        "Description": "example api gateway",
        "Name": "ApiGatewayRestApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ApiGatewayRestApiAccountDF1162EC": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "ApiGatewayRestApi4735CE82",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "ApiGatewayRestApiCloudWatchRole48C066E7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "ApiGatewayRestApiCloudWatchRole48C066E7": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "ApiGatewayRestApiDeploymentEB306364a64ad81357491c8f01d39fcc4f318409": {
      "DependsOn": [
        "ApiGatewayRestApiusersuserIdGETF191F861",
        "ApiGatewayRestApiusersuserIdDB96F9B6",
        "ApiGatewayRestApiusers93E8727D",
      ],
      "Properties": {
        "Description": "example api gateway",
        "RestApiId": {
          "Ref": "ApiGatewayRestApi4735CE82",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiGatewayRestApiDeploymentStageprodFFF8B293": {
      "DependsOn": [
        "ApiGatewayRestApiAccountDF1162EC",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "ApiGatewayRestApiDeploymentEB306364a64ad81357491c8f01d39fcc4f318409",
        },
        "RestApiId": {
          "Ref": "ApiGatewayRestApi4735CE82",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiGatewayRestApiusers93E8727D": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "ApiGatewayRestApi4735CE82",
            "RootResourceId",
          ],
        },
        "PathPart": "users",
        "RestApiId": {
          "Ref": "ApiGatewayRestApi4735CE82",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ApiGatewayRestApiusersuserIdDB96F9B6": {
      "Properties": {
        "ParentId": {
          "Ref": "ApiGatewayRestApiusers93E8727D",
        },
        "PathPart": "{userId}",
        "RestApiId": {
          "Ref": "ApiGatewayRestApi4735CE82",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ApiGatewayRestApiusersuserIdGETApiPermissionRestApiStackApiGatewayRestApi0210B485GETusersuserId247FF107": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "getUserLambda40CC132E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiGatewayRestApi4735CE82",
              },
              "/",
              {
                "Ref": "ApiGatewayRestApiDeploymentStageprodFFF8B293",
              },
              "/GET/users/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiGatewayRestApiusersuserIdGETApiPermissionTestRestApiStackApiGatewayRestApi0210B485GETusersuserId52B860E3": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "getUserLambda40CC132E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ApiGatewayRestApi4735CE82",
              },
              "/test-invoke-stage/GET/users/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiGatewayRestApiusersuserIdGETF191F861": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "getUserLambda40CC132E",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "ApiGatewayRestApiusersuserIdDB96F9B6",
        },
        "RestApiId": {
          "Ref": "ApiGatewayRestApi4735CE82",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiGatewayWafTriggerApiGatewayOpenApiPutFn1601EB8B": {
      "DependsOn": [
        "ApiGatewayWafTriggerApiGatewayOpenApiPutFnServiceRoleDefaultPolicy3235456A",
        "ApiGatewayWafTriggerApiGatewayOpenApiPutFnServiceRoleA483880F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "50d321a0b3483a043ae422562650c90587297d2d44b3dd2d4ebca6f7e3fb29b8.zip",
        },
        "Environment": {
          "Variables": {
            "REST_API_ID": {
              "Ref": "ApiGatewayRestApi4735CE82",
            },
            "STAGE_NAME": "prod",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ApiGatewayWafTriggerApiGatewayOpenApiPutFnServiceRoleA483880F",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ApiGatewayWafTriggerApiGatewayOpenApiPutFnCurrentVersion2B6C2C037407f69838b8df9f1c13dd4fcbbf22a9": {
      "Properties": {
        "FunctionName": {
          "Ref": "ApiGatewayWafTriggerApiGatewayOpenApiPutFn1601EB8B",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "ApiGatewayWafTriggerApiGatewayOpenApiPutFnServiceRoleA483880F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiGatewayWafTriggerApiGatewayOpenApiPutFnServiceRoleDefaultPolicy3235456A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "apigateway:GET",
                "ssm:*",
                "kms:*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ApiGatewayWafTriggerApiGatewayOpenApiPutFnServiceRoleDefaultPolicy3235456A",
        "Roles": [
          {
            "Ref": "ApiGatewayWafTriggerApiGatewayOpenApiPutFnServiceRoleA483880F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ApiGatewayWafTriggerApiGatewayOpenApiPutFnTriggerFF23E50E": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "ApiGatewayRestApiAccountDF1162EC",
        "ApiGatewayRestApiCloudWatchRole48C066E7",
        "ApiGatewayRestApiusersuserIdGETApiPermissionRestApiStackApiGatewayRestApi0210B485GETusersuserId247FF107",
        "ApiGatewayRestApiusersuserIdGETApiPermissionTestRestApiStackApiGatewayRestApi0210B485GETusersuserId52B860E3",
        "ApiGatewayRestApiusersuserIdGETF191F861",
        "ApiGatewayRestApiusersuserIdDB96F9B6",
        "ApiGatewayRestApiusers93E8727D",
        "ApiGatewayRestApiDeploymentEB306364a64ad81357491c8f01d39fcc4f318409",
        "ApiGatewayRestApiDeploymentStageprodFFF8B293",
        "ApiGatewayRestApi4735CE82",
      ],
      "Properties": {
        "HandlerArn": {
          "Ref": "ApiGatewayWafTriggerApiGatewayOpenApiPutFnCurrentVersion2B6C2C037407f69838b8df9f1c13dd4fcbbf22a9",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
            "Arn",
          ],
        },
      },
      "Type": "Custom::Trigger",
      "UpdateReplacePolicy": "Delete",
    },
    "getUserLambda40CC132E": {
      "DependsOn": [
        "getUserLambdaServiceRoleB615547C",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7de530d447468979414e88f458bda8270ccb30a2ce6d9c9cb2a0498e1135108e.zip",
        },
        "Handler": "index.main",
        "Role": {
          "Fn::GetAtt": [
            "getUserLambdaServiceRoleB615547C",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "getUserLambdaServiceRoleB615547C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
